botocore/httpsession.py

```diff
@@ -0,0 +1,279 @@
+    return context
+
+
+class ProxyConfiguration(object):
+    """Represents a proxy configuration dictionary.
+
+    This class represents a proxy configuration dictionary and provides utility
+    functions to retreive well structured proxy urls and proxy headers from the
+    proxy configuration dictionary.
+    """
+    def __init__(self, proxies=None):
+        if proxies is None:
+            proxies = {}
+        self._proxies = proxies
+
+    def proxy_url_for(self, url):
+        """Retrirves the corresponding proxy url for a given url. """
+        parsed_url = urlparse(url)
+        proxy = self._proxies.get(parsed_url.scheme)
+        if proxy:
+            proxy = self._fix_proxy_url(proxy)
+        return proxy
+
+    def proxy_headers_for(self, proxy_url):
+        """Retrirves the corresponding proxy headers for a given proxy url. """
+        headers = {}
+        username, password = self._get_auth_from_url(proxy_url)
+        if username and password:
+            basic_auth = self._construct_basic_auth(username, password)
+            headers['Proxy-Authorization'] = basic_auth
+        return headers
+
+    def _fix_proxy_url(self, proxy_url):
+        if proxy_url.startswith('http:') or proxy_url.startswith('https:'):
+            return proxy_url
+        elif proxy_url.startswith('//'):
+            return 'http:' + proxy_url
+        else:
+            return 'http://' + proxy_url
+
+    def _construct_basic_auth(self, username, password):
+        auth_str = '{0}:{1}'.format(username, password)
+        encoded_str = b64encode(auth_str.encode('ascii')).strip().decode()
+        return 'Basic {0}'.format(encoded_str)
+
+    def _get_auth_from_url(self, url):
+        parsed_url = urlparse(url)
+        try:
+            return unquote(parsed_url.username), unquote(parsed_url.password)
+        except (AttributeError, TypeError):
+            return None, None
+
+
+class URLLib3Session(object):
+    """A basic HTTP client that supports connection pooling and proxies.
+
+    This class is inspired by requests.adapters.HTTPAdapter, but has been
+    boiled down to meet the use cases needed by botocore. For the most part
+    this classes matches the functionality of HTTPAdapter in requests v2.7.0
+    (the same as our vendored version). The only major difference of note is
+    that we currently do not support sending chunked requests. While requests
+    v2.7.0 implemented this themselves, later version urllib3 support this
+    directly via a flag to urlopen so enabling it if needed should be trivial.
+    """
+    def __init__(self,
+                 verify=True,
+                 proxies=None,
+                 timeout=None,
+                 max_pool_connections=MAX_POOL_CONNECTIONS,
+    ):
+        self._verify = verify
+        self._proxy_config = ProxyConfiguration(proxies=proxies)
+        self._pool_classes_by_scheme = {
+            'http': botocore.awsrequest.AWSHTTPConnectionPool,
```

Comment:
It doesn't seem like this change is tested anywhere.  We should test either that we don't mess with the global namespace somehow.

Reply:
Added a test to ensure that the global urllib3 defaults are not set to ours.